### PR TITLE
Player controller Walk & Sprint

### PR DESCRIPTION
**Pull request by Nowp**
It's simply the Walk Controller of base but cleaned and without Duck and Unstuck.

### CODE REVIEW

File: code/Player/Controllers/HumanController.cs [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+		/// <summary>
+		/// This is temporary, get the hull size for the player's collision
+		/// </summary>
+		public override BBox GetHull()
+		{
+			var girth = BodyGirth * 0.5f;
+			var mins = new Vector3( -girth, -girth, 0 );
+			var maxs = new Vector3( +girth, +girth, BodyHeight );
+
+			return new BBox( mins, maxs );
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:41]: What is a BBox ?

File: code/Player/Controllers/HumanController.cs [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+		{
+			var girth = BodyGirth * 0.5f;
+
+			var mins = new Vector3( -girth, -girth, 0 ) * Pawn.Scale;
+			var maxs = new Vector3( +girth, +girth, BodyHeight ) * Pawn.Scale;
+
+			SetBBox( mins, maxs );
+		}
+
+		protected float SurfaceFriction;
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:74]: Why protected prop after methods ?

File: code/Player/Controllers/HumanController.cs [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+
+
+		public override void FrameSimulate()
+		{
+			base.FrameSimulate();
+
+			EyeRot = Input.Rotation;
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:81]: What is the EyeRot ?

File: code/Player/Controllers/HumanController.cs [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+		}
+
+		public override void Simulate()
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:84]: What is the diff between simulate and framesimulate?

File: code/Player/Controllers/HumanController.cs [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+
+			EyePosLocal += TraceOffset;
+			EyeRot = Input.Rotation;
+
+			RestoreGroundPos();
+			
+
+			// RunLadderMode
+
+			CheckLadder();
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:97]: we really want to implement ladder now ?

File: code/Player/Controllers/HumanController.cs [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+			Swimming = Pawn.WaterLevel.Fraction > 0.6f;
+
+			//
+			// Start Gravity
+			//
+			if ( !Swimming && !IsTouchingLadder )
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:103]: We don't need swimming

File: code/Player/Controllers/HumanController.cs [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+			{
+				Velocity -= new Vector3( 0, 0, Gravity * 0.5f ) * Time.Delta;
+				Velocity += new Vector3( 0, 0, BaseVelocity.z ) * Time.Delta;
+
+				BaseVelocity = BaseVelocity.WithZ( 0 );
+			}
+			
+
+			if ( AutoJump ? Input.Down( InputButton.Jump ) : Input.Pressed( InputButton.Jump ) )
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:112]: we don't need autojump, don't use it

File: code/Player/Controllers/HumanController.cs [hunk 8]
```diff
@@ -0,0 +1,676 @@
...
+			}
+
+			//
+			// Work out wish velocity.. just take input, rotate it to view, clamp to -1, 1
+			//
+			WishVelocity = new Vector3( Input.Forward, Input.Left, 0 );
+			var inSpeed = WishVelocity.Length.Clamp( 0, 1 );
+			WishVelocity *= Input.Rotation;
+
+			if ( !Swimming && !IsTouchingLadder )
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:141]: no need swim and ladder

File: code/Player/Controllers/HumanController.cs [hunk 9]
```diff
@@ -0,0 +1,676 @@
...
+			}
+
+			// CheckFalling(); // fall damage etc
+
+			// Land Sound
+			// Swim Sounds
+
+			SaveGroundPos();
+
+			if ( Debug )
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:190]: idk if use debug here is a good practice ? maybe

File: code/Player/Controllers/HumanController.cs [hunk 10]
```diff
@@ -0,0 +1,676 @@
...
+
+			Velocity = Velocity.WithZ( 0 );
+			Accelerate( wishdir, wishspeed, 0, Acceleration );
+			Velocity = Velocity.WithZ( 0 );
+
+			//   Player.SetAnimParam( "forward", Input.Forward );
+			//   Player.SetAnimParam( "sideward", Input.Right );
+			//   Player.SetAnimParam( "wishspeed", wishspeed );
+			//    Player.SetAnimParam( "walkspeed_scale", 2.0f / 190.0f );
+			//   Player.SetAnimParam( "runspeed_scale", 2.0f / 320.0f );
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:232]: dead code

File: code/Player/Controllers/HumanController.cs [hunk 11]
```diff
@@ -0,0 +1,676 @@
...
+				if ( Velocity.Length < 1.0f )
+				{
+					Velocity = Vector3.Zero;
+					return;
+				}
+
+				// first try just moving to the destination
+				var dest = (Position + Velocity * Time.Delta).WithZ( Position.z );
+
+				var pm = TraceBBox( Position, dest );
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:250]: pm is not a comprehensive name

File: code/Player/Controllers/HumanController.cs [hunk 12]
```diff
@@ -0,0 +1,676 @@
...
+			try
+			{
+				if ( Velocity.Length < 1.0f )
+				{
+					Velocity = Vector3.Zero;
+					return;
+				}
+
+				// first try just moving to the destination
+				var dest = (Position + Velocity * Time.Delta).WithZ( Position.z );
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:248]: use var here

File: code/Player/Controllers/HumanController.cs [hunk 13]
```diff
@@ -0,0 +1,676 @@
...
+				Velocity -= BaseVelocity;
+			}
+
+			StayOnGround();
+		}
+
+		public virtual void StepMove()
+		{
+			var startPos = Position;
+			var startVel = Velocity;
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:274]: use var

File: code/Player/Controllers/HumanController.cs [hunk 14]
```diff
@@ -0,0 +1,676 @@
...
+			//
+			// We went further without the step, so lets use that
+			//
+			if ( withoutStep > withStep )
+			{
+				Position = withoutStepPos;
+				Velocity = withoutStepVel;
+				return;
+			}
+		}
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:327]: the method is way too long

File: code/Player/Controllers/HumanController.cs [hunk 15]
```diff
@@ -0,0 +1,676 @@
...
+			// This gets overridden because some games (CSPort) want to allow dead (observer) players
+			// to be able to move around.
+			// if ( !CanAccelerate() )
+			//     return;
+
+			if ( speedLimit > 0 && wishspeed > speedLimit )
+				wishspeed = speedLimit;
+
+			// See if we are changing direction a bit
+			var currentspeed = Velocity.Dot( wishdir );
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:343]: var

File: code/Player/Controllers/HumanController.cs [hunk 16]
```diff
@@ -0,0 +1,676 @@
...
+		{
+			// If we are in water jump cycle, don't apply friction
+			//if ( player->m_flWaterJumpTime )
+			//   return;
+
+			// Not on ground - no friction
+
+
+			// Calculate speed
+			var speed = Velocity.Length;
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:375]: var

File: code/Player/Controllers/HumanController.cs [hunk 17]
```diff
@@ -0,0 +1,676 @@
...
+			bool bMovingUp = Velocity.z > 0;
+
+			bool bMoveToEndPos = false;
+
+			if ( GroundEntity != null ) // and not underwater
+			{
+				bMoveToEndPos = true;
+				point.z -= StepSize;
+			}
+			else if ( bStayOnGround )
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:546]: what means bStayOnGrounbd ? Boddy ?

**Review by Prakkkmak — CHANGES_REQUESTED** [review]
I don't have totally review. I think HumanController is copy pasted. Can you please write the controller yourself with less random stuff. Lots of useless stuff are here. I think you are going to do a much better work with more simple controller for now.
The controller start to be a very big file with responsability issues

File: code/Player/Controllers/HumanController.cs (hunk 1 above)
Comment (Nowp) [reply code/Player/Controllers/HumanController.cs:41]: Bounding Box

File: code/Player/Controllers/HumanController.cs (hunk 4 above)
Comment (Nowp) [reply code/Player/Controllers/HumanController.cs:84]: Simulate = Logic
Frame Simulate = Graphics

File: code/Player/Controllers/HumanController.cs [hunk 18]
```diff
@@ -0,0 +1,676 @@
...
+			//
+			var withoutStepPos = Position;
+			var withoutStepVel = Velocity;
+
+			//
+			// Try again, this time step up and move across
+			//
+			Position = startPos;
+			Velocity = startVel;
+			var trace = TraceBBox( Position, Position + Vector3.Up * (StepSize + DistEpsilon) );
```
[the remark this answers is not in the document]
Comment (Nowp) [reply code/Player/Controllers/HumanController.cs:293]: It's like a raycast but as a box

**Review by Prakkkmak — APPROVED** [review]

**Comment by Prakkkmak** [comment]
[Player controller](https://trello.com/c/gRB9Cqr2/49-player-controller)

File: code/Player/Controllers/HumanController.cs [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+namespace ssl.Player.Controllers
+{
+	public partial class HumanController : BasePlayerController
+	{
+		private Vector3 mins;
+		private Vector3 maxs;
+		private Unstuck unstuck;
+
+		private const float BodyHeight  = 72.0F;
+	    private const float EyeHeight  = 64.0F;
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:14]: weird indent

File: code/Player/Controllers/HumanController.cs [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+			if (IsGrounded)
+			{
+				ApplyFriction(GroundSurface.Friction * SurfaceFriction);
+				Walk();
+			}
+			
+			// DebugOverlay.ScreenText(0, $"    IsGrounded: {IsGrounded}");
+			// DebugOverlay.ScreenText(1,$"       Velocity: {Velocity}  ({Velocity.Length})");
+			// DebugOverlay.ScreenText(2,$"SurfaceFriction: {GroundSurface?.Friction}");
+			// DebugOverlay.ScreenText(3,$"  Ground Normal: {GroundNormal}");
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:63]: remove

File: code/Player/Controllers/HumanController.cs [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+
+		/// <summary>
+		/// Updates the GroundEntity property and all the other related properties like ground normal, surface, etc.
+		/// </summary>
+		private void UpdateGroundEntity()
+		{
+			Vector3 startPos = Position + Vector3.Up * 0.1f;
+			Vector3 endPos = Position - Vector3.Up * 2.0f;
+
+			TraceResult trace = TraceBBox(startPos, endPos, mins, maxs, 4.0F);
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:195]: magic number ?

File: code/Player/Controllers/HumanController.cs [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+				GroundEntity = null;
+				GroundSurface = null;
+				BaseVelocity = Vector3.Zero;
+			}
+		}
+
+		private void Accelerate()
+		{
+			Vector3 acceleration = WishVelocity * Time.Delta;
+
```
Comment (Prakkkmak) [inline code/Player/Controllers/HumanController.cs:216]: ### 
```suggestion
```

File: code/Player/Controllers/HumanController.cs (hunk 19 above)
Comment (Nowp) [reply code/Player/Controllers/HumanController.cs:14]: good on my side

**Review by Prakkkmak — APPROVED** [review]

**Pull request merged**